library/src/tree_node.cpp

```diff
@@ -894,72 +894,127 @@ void CommGather::Print(rocfft_ostream& os, const int indent) const
     if(sendOffsets.size() != num_ranks || sendCounts.size() != num_ranks
        || recvOffsets.size() != num_ranks || recvCounts.size() != num_ranks)
         throw std::runtime_error(
-            "CommAllToAllv: number of counts/offsets does not match number of ranks");
+            "CommAllToAll: number of counts/offsets does not match number of ranks");
 
     if(LOG_PLAN_ENABLED())
     {
-        log_plan("MPI_Ialltoallv\n");
+        log_plan("CommAllToAll: deciding between MPI_Ialltoall and MPI_Ialltoallv\n");
     }
 
 #ifdef ROCFFT_MPI_ENABLE
 
-    // MPI takes ints for everything, convert our size_t elements to int bytes
-    auto convertToInt = [](const std::vector<size_t>& src, std::vector<int>& dest) {
-        dest.reserve(src.size());
-        for(auto i : src)
+    const auto elem_size = element_size(precision, arrayType);
+
+    // check if uniform counts
+    auto count_matches_first = [&](size_t count) { return count == sendCounts[0]; };
+    bool uniform_counts = std::all_of(sendCounts.begin(), sendCounts.end(), count_matches_first)
+                          && std::all_of(recvCounts.begin(), recvCounts.end(), count_matches_first);
+
+    MPI_Request request;
+
+    if(uniform_counts)
+    {
+        if(LOG_PLAN_ENABLED())
+            log_plan("Using MPI_Ialltoall\n");
+
+        if(sendCounts[0] * elem_size > static_cast<size_t>(std::numeric_limits<int>::max()))
         {
-            if(i > std::numeric_limits<int>::max())
-                throw std::runtime_error("MPI integer limit exceeded");
-            dest.push_back(i);
+            std::cerr << "Rank " << local_comm_rank
+                      << ": element size * count_per_rank exceeds MPI_INT limit\n";
+            MPI_Abort(plan->desc.mpi_comm, 1);
         }
-    };
 
-    std::vector<int> intSendOffsets;
-    std::vector<int> intSendCounts;
-    std::vector<int> intRecvOffsets;
-    std::vector<int> intRecvCounts;
-    convertToInt(sendOffsets, intSendOffsets);
-    convertToInt(sendCounts, intSendCounts);
-    convertToInt(recvOffsets, intRecvOffsets);
-    convertToInt(recvCounts, intRecvCounts);
+        const int send_count_bytes = static_cast<int>(sendCounts[0] * elem_size);
+
+        const auto mpiret = MPI_Ialltoall(sendBuf.get(in_buffer, out_buffer, local_comm_rank),
+                                          send_count_bytes,
+                                          MPI_CHAR,
+                                          recvBuf.get(in_buffer, out_buffer, local_comm_rank),
+                                          send_count_bytes,
+                                          MPI_CHAR,
+                                          plan->desc.mpi_comm,
+                                          &request);
+
+        if(mpiret != MPI_SUCCESS)
+        {
+            std::cerr << "MPI_Ialltoall failed on rank " << local_comm_rank
+                      << " with error: " << mpiret << std::endl;
+            MPI_Abort(plan->desc.mpi_comm, mpiret);
```

Comment:
MPI_Abort isn't a solution either.  The collective call has failed, which should mean that it failed on all ranks in the communicator, and now we should return a status code which indicates that things didn't work in the library.  MPI_Abort will exit the process, which stops us from being able to log debug information.

What we should do is get to the collective MPI call, make sure that it fails, then return an error status code.